File: booking_service/src/core/usecase/get_accommodation_usecase.go
package usecase

import (
	"booking_service/core/domain/constant"
	"booking_service/core/domain/entity"
	"booking_service/core/port"
	"booking_service/kernel/utils"
	"context"
	"encoding/json"
	"github.com/golibs-starter/golib/log"
)

type IGetAccommodationUseCase interface {
	GetAccommodationByID(ctx context.Context, ID int64) (*entity.AccommodationEntity, error)
}

type GetAccommodationUseCase struct {
	accPort   port.IAccommodationPort
	unitPort  port.IUnitPort
	cachePort port.ICachePort
}

func (g GetAccommodationUseCase) GetAccommodationByID(ctx context.Context, ID int64) (*entity.AccommodationEntity, error) {
	// get from cache
	cachedAcc, err := g.cachePort.GetFromCache(ctx, utils.BuildCacheKeyGetAccommodation(ID))
	if err != nil && err.Error() != constant.ErrCacheKeyNil {
		log.Error(ctx, "Get AccommodationByID failed", err)
		return nil, err
	}
	if cachedAcc != nil {
		response := entity.AccommodationEntity{}
		err = json.Unmarshal([]byte(cachedAcc.(string)), &response)
		if err != nil {
			log.Error(ctx, "Unmarshal error", err)
			return nil, err
		}
		return &response, nil
	}

	acc, err := g.accPort.GetAccommodationByID(ctx, ID)
	if err != nil {
		log.Error(ctx, "get accommodation by id failed", err)
		return nil, err
	}

	units, err := g.unitPort.GetUnitsByAccID(ctx, acc.ID)
	if err != nil {
		log.Error(ctx, "get units failed", err)
		return nil, err
	}
	acc.Units = units

	go func() {
		err = g.cachePort.SetToCache(ctx, utils.BuildCacheKeyGetAccommodation(acc.ID), acc, constant.DefaultCacheTTL)
		if err != nil {
			log.Error(ctx, "SetToCache error", err)
		}
	}()

	return acc, nil
}

func NewGetAccommodationUseCase(accPort port.IAccommodationPort, unitPort port.IUnitPort, cachePort port.ICachePort) IGetAccommodationUseCase {
	return &GetAccommodationUseCase{accPort: accPort, unitPort: unitPort, cachePort: cachePort}
}

File: inventory_service/src/main/java/huy/project/inventory_service/core/usecase/GetAccommodationUseCase.java
package huy.project.inventory_service.core.usecase;

import huy.project.inventory_service.core.domain.constant.CacheConstant;
import huy.project.inventory_service.core.domain.entity.Accommodation;
import huy.project.inventory_service.core.domain.exception.NotFoundException;
import huy.project.inventory_service.core.port.IAccommodationPort;
import huy.project.inventory_service.core.port.ICachePort;
import huy.project.inventory_service.kernel.util.CacheUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class GetAccommodationUseCase {
    private final IAccommodationPort accommodationPort;
    private final ICachePort cachePort;
    private final GetUnitUseCase getUnitUseCase;

    public Accommodation getAccommodationById(Long id) {
        return getAccommodationById(id, false);
    }

    public Accommodation getAccommodationById(Long id, boolean includeUnits) {
        var cacheKey = CacheUtils.buildCacheKeyGetAccById(id, includeUnits);
        var cachedAcc = cachePort.getFromCache(cacheKey, Accommodation.class);
        if (cachedAcc != null) {
            return cachedAcc;
        }

        var accommodation = accommodationPort.getAccommodationById(id);
        if (accommodation == null) {
            throw new NotFoundException("Accommodation not found, id: " + id);
        }

        if (includeUnits) {
            accommodation.setUnits(getUnitUseCase.getUnitsByAccommodationId(id, true));
        }

        cachePort.setToCache(cacheKey, accommodation, CacheConstant.DEFAULT_TTL);

        return accommodation;
    }
}
